File: Python/fileio.py
#!/usr/bin/env python3
import time
import os

ITERATIONS = 1000
FILE_SIZE = 5000000


def main():
    print(f"Python File I/O Benchmark ({ITERATIONS} iterations)")

    data = bytes([ord("a") + (i % 26) for i in range(FILE_SIZE)])

    if os.path.exists("/tmp/bench_io.dat"):
        os.remove("/tmp/bench_io.dat")

    t0 = time.perf_counter()
    for _ in range(ITERATIONS):
        with open("/tmp/bench_io.dat", "wb") as f:
            f.write(data)
    t1 = time.perf_counter()
    ms = (t1 - t0) * 1000
    print(f"Write: {ms:.2f} ms ({ITERATIONS / (ms / 1000):.0f} ops/sec)")

    t0 = time.perf_counter()
    for _ in range(ITERATIONS):
        with open("/tmp/bench_io.dat", "rb") as f:
            _ = f.read()
    t1 = time.perf_counter()
    ms = (t1 - t0) * 1000
    print(f"Read: {ms:.2f} ms ({ITERATIONS / (ms / 1000):.0f} ops/sec)")

    t0 = time.perf_counter()
    for _ in range(ITERATIONS):
        with open("/tmp/bench_io.dat", "r") as f:
            for line in f:
                pass
    t1 = time.perf_counter()
    ms = (t1 - t0) * 1000
    print(f"Read lines: {ms:.2f} ms ({ITERATIONS / (ms / 1000):.0f} ops/sec)")

    t0 = time.perf_counter()
    for _ in range(ITERATIONS):
        with open("/tmp/bench_io.dat", "rb") as f:
            f.seek(FILE_SIZE // 2)
            _ = f.read(1024)
    t1 = time.perf_counter()
    ms = (t1 - t0) * 1000
    print(f"Random access: {ms:.2f} ms ({ITERATIONS / (ms / 1000):.0f} ops/sec)")

    if os.path.exists("/tmp/bench_io.dat"):
        os.remove("/tmp/bench_io.dat")


if __name__ == "__main__":
    main()

File: Rust/fileio.rs
use std::fs::{File};
use std::io::{Read, Write, BufRead, BufReader, Seek, SeekFrom};
use std::time::Instant;

const ITERATIONS: usize = 1000;
const FILE_SIZE: usize = 5000000;

fn main() {
    println!("Rust File I/O Benchmark ({} iterations)", ITERATIONS);
    
    let data: Vec<u8> = (0..FILE_SIZE).map(|i| b'a' + (i % 26) as u8).collect();
    
    let _ = std::fs::remove_file("/tmp/bench_io.dat");
    
    let t0 = Instant::now();
    for _ in 0..ITERATIONS {
        let mut f = File::create("/tmp/bench_io.dat").unwrap();
        f.write_all(&data).unwrap();
    }
    let t1 = Instant::now();
    let ms = (t1 - t0).as_secs_f64() * 1000.0;
    println!("Write: {:.2} ms ({:.0} ops/sec)", 
        ms, ITERATIONS as f64 / (ms / 1000.0));
    
    let t0 = Instant::now();
    for _ in 0..ITERATIONS {
        let mut f = File::open("/tmp/bench_io.dat").unwrap();
        let mut buf = vec![0u8; FILE_SIZE];
        f.read_exact(&mut buf).unwrap();
    }
    let t1 = Instant::now();
    let ms = (t1 - t0).as_secs_f64() * 1000.0;
    println!("Read: {:.2} ms ({:.0} ops/sec)", 
        ms, ITERATIONS as f64 / (ms / 1000.0));
    
    let t0 = Instant::now();
    for _ in 0..ITERATIONS {
        let f = File::open("/tmp/bench_io.dat").unwrap();
        let reader = BufReader::new(f);
        for line in reader.lines() {
            let _ = line;
        }
    }
    let t1 = Instant::now();
    let ms = (t1 - t0).as_secs_f64() * 1000.0;
    println!("Read lines: {:.2} ms ({:.0} ops/sec)", 
        ms, ITERATIONS as f64 / (ms / 1000.0));
    
    let t0 = Instant::now();
    for _ in 0..ITERATIONS {
        let mut f = File::open("/tmp/bench_io.dat").unwrap();
        f.seek(SeekFrom::Start((FILE_SIZE / 2) as u64)).unwrap();
        let mut buf = vec![0u8; 1024];
        f.read(&mut buf).unwrap();
    }
    let t1 = Instant::now();
    let ms = (t1 - t0).as_secs_f64() * 1000.0;
    println!("Random access: {:.2} ms ({:.0} ops/sec)", 
        ms, ITERATIONS as f64 / (ms / 1000.0));
    
    let _ = std::fs::remove_file("/tmp/bench_io.dat");
}
